### PR TITLE
Fixed #32348 -- Made deleting "extra" inlines in admin impossible.

### PR DESCRIPTION
**Pull request by hramezani**
https://code.djangoproject.com/ticket/32348

### CODE REVIEW

**Comment by hramezani** [comment]
The second commit is from @youguanxinqing. Here is the [detail](https://code.djangoproject.com/ticket/32348#comment:8).

File: django/contrib/admin/static/admin/js/inlines.js [hunk 1]
```diff
@@ -138,7 +140,25 @@
             if ((maxForms.val() === '') || (maxForms.val() - forms.length) > 0) {
                 addButton.parent().show();
             }
-            // Hide the remove buttons if at min_num.
+
+            const isFormNotEnough = function() {
+                const totalFormNum = ~~totalForms.val();
```
Comment (carltongibson) [inline django/contrib/admin/static/admin/js/inlines.js:145]: Here, we're worried about coverting to an int right? Why not `parseInt()`? (Is that not clearer in intent?)

File: django/forms/formsets.py [hunk 2]
```diff
@@ -34,6 +35,7 @@ class ManagementForm(Form):
     def __init__(self, *args, **kwargs):
         self.base_fields[TOTAL_FORM_COUNT] = IntegerField(widget=HiddenInput)
         self.base_fields[INITIAL_FORM_COUNT] = IntegerField(widget=HiddenInput)
+        self.base_fields[EXTRA_FORM_COUNT] = IntegerField(required=False, widget=HiddenInput)
```
Comment (carltongibson) [inline django/forms/formsets.py:38]: Do we need the extra field here? Could we do something like this?

`EXTRA_FORM_COUNT  = TOTAL_FORM_COUNT - max(INITIAL_FORM_COUNT, MIN_NUM_FORM_COUNT)` 

(c.f. https://github.com/django/django/blob/d64f192df188903ea5a1ab8dfd6679e5d52aaf9b/django/forms/formsets.py#L117)

Either way `EXTRA_FORM_COUNT ` is purely for client code, yes? So it should go beneath the comment below. The comment should be adjusted, along with [the note on Understanding the management form](https://github.com/django/django/blob/d3b3eb860dc140c56022704672e7f54fa8a0c648/docs/topics/forms/formsets.txt#L262-L269) in the topic docs. 

(Sure we can't calculate it?)

File: django/contrib/admin/static/admin/js/inlines.js [hunk 3]
```diff
@@ -153,11 +173,41 @@
...
+                const initialFormNum = ~~initialForms.val();
+                const minFormNum = ~~minForms.val();
+                const extraFormNum = ~~extraForms.val();
+
+                const emptyFormNum = totalFormNum - initialFormNum;
+                if (initialFormNum >= emptyFormNum) {
+                    return emptyFormNum > extraFormNum;
+                }
+                return emptyFormNum > extraFormNum + minFormNum - initialFormNum;
+            }
```
Comment (carltongibson) [inline django/contrib/admin/static/admin/js/inlines.js:197]: Can we refactor the duplicate block?

**Review by carltongibson — COMMENTED** [review]
Hi @hramezani and @youguanxinqing — thanks for this. 

Looks along the right lines. I have a couple of initial questions? 

Needs a release note in `3.1.6.txt`

**Comment by hramezani** [comment]
Hi @carltongibson,

Thanks for the review.

As you suggested I tried to calculate `extraForms` in `inlines.js` file. So I removed the extra integer field from in formset.
Also, I removed the extra function in `inlines.js` which was added in the previous commit.

We still have one failing test 
https://github.com/django/django/blob/f054468cac325e8d8fa4d5934b939b93242a3730/js_tests/admin/inlines.test.js#L171

I don't know should we make visible the delete button for a formset with a validation error?

Also, I will add a release note soon

**Comment by carltongibson** [comment]
Hey @hramezani, super! 

> As you suggested I tried to calculate extraForms in inlines.js file. So I removed the extra integer field from in formset.

This worked as expected in your view? (I found myself wondering if there were edge cases... 🙂)

I'll take another look tomorrow. Thanks for your effort! 👍

File: django/contrib/admin/static/admin/js/inlines.js (hunk 1 above)
Comment (youguanxinqing) [reply django/contrib/admin/static/admin/js/inlines.js]: > Here, we're worried about coverting to an int right? Why not `parseInt()`? (Is that not clearer in intent?)

`~~`  could also convert string to int and have high fault tolerance. Of, course, `parseInt()`  and `~~` both fit here and the former is more readable.

**Comment by youguanxinqing** [comment]
> I don't know should we make visible the delete button for a formset with a validation error?

see [ticket 29087](https://code.djangoproject.com/ticket/29087): **Added delete buttons for unsaved admin inlines on validation error.**

File: django/contrib/admin/static/admin/js/inlines.js [hunk 4]
```diff
@@ -140,7 +140,25 @@
...
+                if (initialFormNum >= emptyFormNum) {
+                    return emptyFormNum < extraFormNum;
+                }
+                return emptyFormNum < extraFormNum + minFormNum - initialFormNum;
+            }
+            // Auto add form when forms are not enough
+            if (isFormNotEnough()) {
+                $('.' + options.addCssClass +' a').click();
+            }
+
```
Comment (youguanxinqing) [inline django/contrib/admin/static/admin/js/inlines.js]: Sorry, why delete the logic here? The failure of test is probably related to it. 

(I found I have no permission to push code to the pr, I couldn't chang anything.

File: django/forms/formsets.py (hunk 2 above)
Comment (youguanxinqing) [reply django/forms/formsets.py]: `extra` filed is assigned by user. for example:
```python
class ChoiceInline(admin.StackedInline):
    model = Choice
    extra = 3
```
ref: https://docs.djangoproject.com/en/3.1/intro/tutorial07/#writing-your-first-django-app-part-7

I'am sorry that I think `EXTRA_FORM_COUNT = TOTAL_FORM_COUNT - max(INITIAL_FORM_COUNT, MIN_NUM_FORM_COUNT)` is incorrect.

Comment (carltongibson) [reply django/forms/formsets.py]: Hi @youguanxinqing — but `total_forms` is basically just `extra` + the initial forms, so if we have those two we _shouldn't_ need to set it manually (is my hunch). 

Maybe there's an edge case but I'd like to see that, and have it covered by a test if so. 
I'll have a play with it now, but if you can show how it's wrong, that would be great! 
(Thanks! 👍)

**Comment by felixxm** [comment]
Unfortunately, it doesn't work for me. In the first step `delete` links are not visible (that's good):
image
but when I click `Add another Choice` they appear for all choices:
image

**Comment by hramezani** [comment]
> but when I click `Add another Choice` they appear for all choices:

Shouldn't it be like this? Delete links appear for all choices and after you clicked on one of those(delete on extra item), they become invisible again.

**Comment by felixxm** [comment]
> Shouldn't it be like this? Delete links appear for all choices and after you clicked on one of those(delete on extra item), they become invisible again.

Previously, it was only visible for manually added choices.

**Comment by hramezani** [comment]
> > Shouldn't it be like this? Delete links appear for all choices and after you clicked on one of those(delete on extra item), they become invisible again.
> 
> Previously, it was only visible for manually added choices.

Based on this line of code 
https://github.com/django/django/blob/0c42cdf0d2422f4c080e93594d5d15381d6e955e/django/contrib/admin/static/admin/js/inlines.js#L159
It seems that it will be visible now for all choices. 
Do you mean that the current implementation is wrong?

File: django/forms/formsets.py (hunk 2 above)
Comment (youguanxinqing) [reply django/forms/formsets.py]: Sorry for my mistabke, you are right.

**Comment by carltongibson** [comment]
> Do you mean that the current implementation is wrong?

Hey @hramezani — yes, it looks that way — It's simply not taking into account the original `extra` forms. 

Those should not have the (X), unless a validation error has occurred — in which case, as @youguanxinqing says, the point of ticket-29087 was to make clearing them easier. 

I had a bit of a play this afternoon, and it gets fiddly trying to do too much reading state from the number of DOM elements the whole time — but I'm reluctant to suggest too much as a backport. 🤔 

Reviewing https://github.com/django/django/pull/11504 it might be that a small adjustment to the tutorial is in order, perhaps in combination with a small change here. (I'm not 100% sure yet what the best approach is.)

**Comment by carltongibson** [comment]
Hi all, right... sorry... Reviewing #11504 (again), the change in behaviour here was deliberate. 

> With this PR, those 3 inlines will have the delete buttons. Every inline will have either the delete button (for unsaved forms) or the delete checkbox (for saved forms). Permissions are respected as before.

Thinking that through (again), I (again) agree with the reasoning there that the change makes sense. (i.e. it's a desired change)

What's happened is that we've failed to update the tutorial documentation to match. 
I'll double-check with @felixxm about that conclusion, but I think it's a docs change that's needed. 

Thanks for the hard work!

**Comment by carltongibson** [comment]
Suggested an alternative in #13936

**Comment by carltongibson** [comment]
Having conferred with @felixxm, closing in favour of #13936. Thanks both!

**Pull request closed without merging**